contracts/cw-escrow/src/contract.rs

```diff
@@ -0,0 +1,225 @@
+        storage.save(deps.storage, &counterparty)?;
+
+        let messages = if counterparty.provided && other_counterparty.provided {
+            vec![
+                counterparty
+                    .promise
+                    .into_send_message(&other_counterparty.address)?,
+                other_counterparty
+                    .promise
+                    .into_send_message(&counterparty.address)?,
+            ]
+        } else {
+            vec![]
+        };
+
+        Ok(Response::new()
+            .add_attribute("method", "receive_cw20")
+            .add_attribute("counterparty", counterparty.address)
+            .add_messages(messages))
+    } else {
+        Err(ContractError::InvalidFunds {})
+    }
+}
+
+pub fn execute_fund(deps: DepsMut, info: MessageInfo) -> Result<Response, ContractError> {
+    let counterparty_one = COUNTERPARTY_ONE.load(deps.storage)?;
+    let counterparty_two = COUNTERPARTY_TWO.load(deps.storage)?;
+
+    let (counterparty, other_counterparty, storage) = if info.sender == counterparty_one.address {
+        (counterparty_one, counterparty_two, COUNTERPARTY_ONE)
+    } else if info.sender == counterparty_two.address {
+        (counterparty_two, counterparty_one, COUNTERPARTY_TWO)
+    } else {
+        // Contract may only be funded by a counterparty.
+        return Err(ContractError::Unauthorized {});
+    };
+
+    if counterparty.provided {
+        return Err(ContractError::AlreadyProvided {});
+    }
+
+    if let CheckedTokenInfo::Native { amount, denom } = &counterparty.promise {
+        let paid = must_pay(&info, denom).map_err(|_| ContractError::InvalidFunds {})?;
+        if paid != *amount {
+            return Err(ContractError::InvalidAmount {
+                expected: *amount,
+                actual: paid,
+            });
+        }
+
+        let mut counterparty = counterparty;
+        counterparty.provided = true;
+        storage.save(deps.storage, &counterparty)?;
+
+        let messages = if counterparty.provided && other_counterparty.provided {
+            vec![
+                counterparty
+                    .promise
+                    .into_send_message(&other_counterparty.address)?,
+                other_counterparty
+                    .promise
+                    .into_send_message(&counterparty.address)?,
+            ]
+        } else {
+            vec![]
+        };
+
+        Ok(Response::new()
+            .add_attribute("method", "fund")
+            .add_attribute("counterparty", counterparty.address)
+            .add_messages(messages))
+    } else {
+        Err(ContractError::InvalidFunds {})
+    }
+}
+
+pub fn execute_withdraw(deps: DepsMut, info: MessageInfo) -> Result<Response, ContractError> {
+    let counterparty_one = COUNTERPARTY_ONE.load(deps.storage)?;
+    let counterparty_two = COUNTERPARTY_TWO.load(deps.storage)?;
+
+    let (counterparty, other_counterparty, storage) = if info.sender == counterparty_one.address {
```

Comment:
lets make this a function

Reply:
Great suggestion. See `get_counterparty` in the most recent commit. :)